lib/ddtrace/contrib/sinatra/framework.rb

```diff
@@ -0,0 +1,77 @@
+module Datadog
+  module Contrib
+    # Instrument Sinatra.
+    module Sinatra
+      # Sinatra framework code, used to essentially:
+      # - handle configuration entries which are specific to Datadog tracing
+      # - instrument parts of the framework when needed
+      module Framework
+        # Configure Rack from Sinatra, but only if Rack has not been configured manually beforehand
+        def self.setup
+          return if Datadog::Tracing.configuration.instrumented_integrations.key?(:rack)
+
+          Datadog::Tracing.configure do |datadog_config|
+            sinatra_config = config_with_defaults(datadog_config)
+            activate_rack!(datadog_config, sinatra_config)
+          end
+        end
+
+        def self.config_with_defaults(datadog_config)
+          datadog_config[:sinatra]
+        end
+
+        # Apply relevant configuration from Sinatra to Rack
+        def self.activate_rack!(datadog_config, sinatra_config)
+          datadog_config.instrument(
+            :rack,
+            service_name: sinatra_config[:service_name],
```

Comment:
Should Rack receive Sinatra's service name? Generally speaking, I don't think this is an issue when in most cases we want all instrumentation to reflect the same service name. However, this will not allow users to override Rack service names unless they do it after this configuration occurs.

Reply:
This is consistent with the Rails instrumentation today: https://github.com/DataDog/dd-trace-rb/blob/93cd757770df59455fc94d04bb2db62f3dd22ada/lib/ddtrace/contrib/rails/framework.rb#L73-L76

It seems harsh to not allow `c.use :rack, service_name:` to take place, but it's what Rails does.

There's an argument that allowing two different service names between Rack and Rails/Sinatra is asking for confusion.

TBH, I believe we are moving towards removing the ability to configure custom service names for internal integrations separately, so having them have the same service name is, in a roundabout way, aligned with our simplified service naming goals.

Reply:
Indeed as a general trend in this PR I tried to port over how the Rails integration behaves, taking it as a baseline for behaviour.

It is arguable whether Rack should be considered an independent service or part of the service that depends on it (Sinatra, Rails), but I would argue for the latter, not the least because the frameworks pick specific Rack middlewares from the `rack` gem and interleave these with their own, so the framework implements itself not just "over"/"after" Rack but "with" Rack.

Reply:
> I believe we are moving towards removing the ability to configure custom service names for internal integrations separately

That was the impetus for the guard glause, which according to what you say, might be irrelevant:

```
return if Datadog::Tracing.configuration.instrumented_integrations.key?(:rack)
```

Come to think of it, in the case of nested apps, I wonder how configuring the Rack integration service name would be expected to work, if at all (I didn't check but intuitively it seems all Rack services would be named the same, nested or not?)